boot/freeldr/freeldr/ntldr/arch/i386/winldr.c

```diff
@@ -186,33 +184,46 @@ MempAllocatePageTables(VOID)
 {
     ULONG NumPageTables, TotalSize;
     PUCHAR Buffer;
-    // It's better to allocate PDE + PTEs contiguous
+    ULONG ix;
 
-    // Max number of entries = MaxPageNum >> 10
-    // FIXME: This is a number to describe ALL physical memory
-    // and windows doesn't expect ALL memory mapped...
-    NumPageTables = TotalPagesInLookupTable >> 10;
+    TRACE("MempAllocatePageTables: PaeModeOn %d\n", PaeModeOn);
 
-    TRACE("NumPageTables = %d\n", NumPageTables);
+    /* It's better to allocate PDE + PTEs contiguous.
+       Max number of entries = MaxPageNum >> 10(9 if PAE).
+       FIXME: This is a number to describe ALL physical memory
+       and windows doesn't expect ALL memory mapped...
+    */
 
-    // Allocate memory block for all these things:
-    // PDE, HAL mapping page table, physical mapping, kernel mapping
-    TotalSize = (1 + 1 + NumPageTables * 2) * MM_PAGE_SIZE;
+    /* Allocate memory block for all these things:
+       PDE, HAL mapping page table, physical mapping, kernel mapping.
+    */
+    if (!PaeModeOn)
+    {
+        NumPageTables = (TotalPagesInLookupTable / PTE_PER_PAGE);
+        TotalSize = ((1 + 1 + (NumPageTables * 2)) * MM_PAGE_SIZE);
+        TRACE("MempAllocatePageTables: NumPageTables = %X (%d)\n", NumPageTables, NumPageTables);
+    }
+    else
+    {
+       NumPageTables = (TotalPagesInLookupTable / PAE_PTE_PER_PAGE);
+       TotalSize = (sizeof(PAE_TABLES) + ((NumPageTables * 2) * MM_PAGE_SIZE));
+        TRACE("MempAllocatePageTables: NumPageTables = %X (%d)\n", NumPageTables, NumPageTables);
+    }
```

Comment:
Moving common trace to both if-clauses outside:
```suggestion
    }
    TRACE("MempAllocatePageTables: NumPageTables = %X (%d)\n", NumPageTables, NumPageTables);
```

Reply:
Ok